### PR TITLE
quiche: drain quic listener

### LINKED ISSUE #7465: network: UdpListener stops abruptly
ConnectionHandlerImpl::stopListeners() destroy the listener instance immediately to stop listening. This works for TcpListener which only listens for new connections, and existing connections have their own sockets registered in scheduler. But UdpListener has only one socket for all the I/O. If the UdpListener was destroyed, the existing UDP traffic will just stop. It is fine if the UDP traffic is connection-less. But if it's connection oriented, all the existing connections have no way to continue. This behavior is different from the intention of stopListeners(). 

We need to refactor ConnectionHandler and UdpListener to support draining feature. A rough idea is to add an interface ActiveListenerBase::stopListening() with a callback which actually destroy the listener. And stopListening() should
1. For ActiveTcpListener, call the callback immediately.
2. For UdpListenerCallback, tell ListenerFilter to stop listening according the the filters own definition, and call the callback when the filter thinks it finishes draining.

### PR DESCRIPTION
**Pull request by danzh2010**
Rename ActiveListenerBase::destroy() to shutdownListener(). Add two more interfaces: pauseListening() and resumeListening() which in ActiveTcpListener wraps listener_->disable() and enable(), and in ActiveQuicListener wraps QuicDispatcher::StopAcceptingNewConnections() and StartAcceptingNewConnections(). ActiveListenerBase::pause/resumeListening() is used in draining listeners.

Risk Level: low
Testing: added admin draning test and overload test
Fixes #7465
Part of #2557

### CODE REVIEW

File: include/envoy/network/connection_handler.h [hunk 1]
```diff
@@ -97,9 +97,19 @@ class ConnectionHandler {
...
+
+    /**
+     * Resume listening according to implementation's own definition.
+     */
+    virtual void resumeListening() PURE;
+
+    /**
+     * Stop listening according to implementation's own definition.
+     */
+    virtual void stopListening() PURE;
```
Comment (mattklein123) [inline include/envoy/network/connection_handler.h]: `pause` vs. `stop` is kind of confusing. Should the existing `stopListening` become `destroy()`? `shutdown()`? WDYT?

Comment (danzh2010) [reply include/envoy/network/connection_handler.h]: Do you mean ConnectionHandler::stopListeners()? I'm not good at naming at all. IMO what the existing stopListeners() really means is stop accepting. The TCP implementation is calling the existing destroy() which makes sense as a TCP listener only listen on new connection. But the UDP implementation stop accepting doesn't mean destroying the listener. That's why I renamed destroy() to stopListening() and let TCP and UDP implementation divert. The existing stopListeners() is an appropriate name for TCP but not for UDP.

I agree that stop and pause aren't informative enough. I named them so to indicate the difference is the former is irreversible and the latter is reversible. This is corresponding to ConnectionHandler's stopListeners() and enable|disableListeners().  Do you have better naming strategy for them?

Comment (mattklein123) [reply include/envoy/network/connection_handler.h]: How about `shutdownListening()`? It's minor but I think it makes it more clear. And then int he doc comments maybe more clearly indicate the difference between shutdown/pause? WDYT?

**Comment by danzh2010** [comment]
The coverage ci failure seems not related..

**Pull request merged**